File: apps/atlas-api/src/features/plan-notices/presentation/http/handle-plan-notices.ts
import type { RuntimeContext } from "../../../../core/runtime-context";
import type { Telemetry } from "../../../telemetry/application/ports/telemetry";
import type { IdentityStore } from "../../../identity/application/ports/identity-store";
import { fw } from "../../../../../fireweave/fw-harness";
import { getPlanNotices } from "../../application/get-plan-notices";

function json(data: unknown, status = 200): Response {
  return new Response(JSON.stringify(data) + "\n", {
    status,
    headers: { "content-type": "application/json" },
  });
}

function bearer(req: Request): string | undefined {
  const h = req.headers.get("authorization") ?? "";
  const m = /^Bearer\s+(.+)$/i.exec(h);
  return m?.[1]?.trim();
}

export async function handlePlanNotices(
  ctx: RuntimeContext,
  telemetry: Telemetry,
  identity: IdentityStore,
  req: Request,
): Promise<Response> {
  const started = performance.now();
  const sess = identity.session(bearer(req));
  if (!sess) {
    return json({ error: "unauthenticated" }, 401);
  }

  // @fireweave-controlpoint plan-notices
  const enabled = await fw.controlPoints.getBooleanValue("plan-notices", false, {
    targetingKey: sess.user.id,
  });
  if (!enabled) {
    return json({ error: "not found" }, 404);
  }

  try {
    const body = getPlanNotices(sess.user);
    const elapsed = performance.now() - started;
    telemetry.record("plan.notices.latency_ms", elapsed, {
      plan: sess.user.plan,
    });
    return json(body);
  } catch {
    telemetry.increment("plan.notices.request.server_error", {
      plan: sess.user.plan,
    });
    return json({ error: "notices unavailable" }, 500);
  }
}

File: apps/lantern-api/src/features/plan_notices/presentation/http/handle_plan_notices.py
import json
import re
import time
from typing import Optional

from fireweave._sdk import EvaluationContext

from fireweave.fw_harness import fw_control_points
from src.features.identity.domain.user import BenchUser
from src.features.plan_notices.application.get_plan_notices import get_plan_notices


def _json(data: dict, status: int = 200) -> tuple[int, bytes]:
    return status, (json.dumps(data) + "\n").encode()


def _bearer(headers) -> Optional[str]:
    auth = headers.get("Authorization") or headers.get("authorization") or ""
    m = re.match(r"^Bearer\s+(.+)$", auth, re.I)
    return m.group(1).strip() if m else None


def handle_plan_notices(identity, telemetry, headers) -> tuple[int, bytes]:
    started = time.perf_counter()
    sess = identity.session(_bearer(headers))
    if not sess:
        return _json({"error": "unauthenticated"}, 401)

    user_row = sess["user"]
    user = BenchUser(**user_row)
    plan = user.plan

    # @fireweave-controlpoint plan-notices
    enabled = fw_control_points().get_boolean_value(
        "plan-notices",
        False,
        EvaluationContext(targeting_key=user.id),
    )
    if not enabled:
        return _json({"error": "not found"}, 404)

    try:
        body = get_plan_notices(user)
        elapsed_ms = (time.perf_counter() - started) * 1000
        telemetry.record(
            "plan.notices.latency_ms",
            elapsed_ms,
            attrs={"plan": plan},
        )
        return _json(body)
    except Exception:
        telemetry.increment(
            "plan.notices.request.server_error",
            attrs={"plan": plan},
        )
        return _json({"error": "notices unavailable"}, 500)
